### PR TITLE
Asynchronous search namespace

### PR DESCRIPTION
**Pull request by Tokesh**
### Description
`asynchronous_search` namespace

### Issues Resolved
[#223]

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/test-spec.yml [hunk 1]
```diff
@@ -73,7 +73,7 @@ jobs:
       - name: Run Tests
         run: |
           npm run test:spec -- \
-            --opensearch-insecure \
+            --opensearch-insecure --verbose \
```
Comment (dblock) [inline .github/workflows/test-spec.yml:76]: Let's either remove `--verbose` or put it on its own line.

File: CHANGELOG.md [hunk 2]
```diff
@@ -79,6 +79,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Added `cancelled_task_percentage` and `current_cancellation_eligible_tasks_count` to `ShardSearchBackpressureTaskCancellationStats` ([#483](https://github.com/opensearch-project/opensearch-api-specification/pull/483))
 - Added detailed test coverage report ([#513](https://github.com/opensearch-project/opensearch-api-specification/pull/513))
 - Added 404 responses to `/_alias/{name}` and `/{index}/_alias/{name}` ([#519](https://github.com/opensearch-project/opensearch-api-specification/pull/519))
+- Added `asynchronous_search` namespace API specifications ([#525](https://github.com/opensearch-project/opensearch-api-specification/pull/525))
```
Comment (dblock) [inline CHANGELOG.md:82]:
```suggestion
- Added `asynchronous_search` ([#525](https://github.com/opensearch-project/opensearch-api-specification/pull/525))
```

File: tests/default/asynchronous_search/search.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  - path: /_bulk
+    method: POST
+    request:
+      content_type: application/x-ndjson
+      payload:
+        - {create: {_index: books, _id: book_1392214}}
+        - {author: Harper Lee, title: To Kill a Mockingbird, year: 1960}
+        - {create: {_index: books, _id: book_1392215}}
+        - {author: Elizabeth Rudnick, title: Beauty and the Beast, year: 1991}
+  - path: /{index}/_refresh
```
Comment (dblock) [inline tests/default/asynchronous_search/search.yaml:21]: Nit: I think `_bulk` supports `refresh` as well if you want to collapse.

File: tests/default/asynchronous_search/search.yaml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      payload:
+        query:
+          match:
+            author:
+              query: Bennett Miller
+    output:
+      id: payload.id
+    response:
+      status: 200
+  - synopsis: Get partial response from asynchronous search.
```
Comment (dblock) [inline tests/default/asynchronous_search/search.yaml:47]: So is it possible it hasn't completed yet (are we introducing a flaky test)?

**Review by dblock — CHANGES_REQUESTED** [review]
Looks good, I have a few nits.

**Comment by Tokesh** [comment]
Hello everyone! 
I added asynchronous search to the specs, wrote a couple of tests. I noticed that in OS 3.0 they do not pass the tests, can you tell me why this might be. With other versions these tests pass successfully
I didn't find a word about removing asynchronous search in OS 3.0.
@dblock @nhtruong

**Comment by dblock** [comment]
> Hello everyone! I added asynchronous search to the specs, wrote a couple of tests. I noticed that in OS 3.0 they do not pass the tests, can you tell me why this might be. With other versions these tests pass successfully I didn't find a word about removing asynchronous search in OS 3.0. @dblock @nhtruong

It looks like asynchronous search tests failed with the following:

```

[INFO] => POST /_plugins/_asynchronous_search ({
  "index": "books",
  "keep_on_completion": true
}) [application/json] {
  "query": {
    "match": {
      "author": {
        "query": "Bennett Miller"
      }
    }
  }
}
[INFO] <= 400 (application/json) | "no handler found for uri [/_plugins/_asynchronous_search] and method [POST]"
[INFO] => DELETE /books ({}) [application/json] 
[INFO] <= 200 (application/json; charset=UTF-8) | {
  "acknowledged": true
}
```

This means that the plugin is not in this 3.0 build (3.0 builds can be partial). Look for a newer tag on https://hub.docker.com/r/opensearchstaging/opensearch/tags that includes the asynchronous-search plugin as well and replace it in CI (might need to try locally).

**Comment by Tokesh** [comment]
> This means that the plugin is not in this 3.0 build (3.0 builds can be partial). Look for a [newer tag](https://hub.docker.com/layers/opensearchstaging/opensearch/3.0.0/images/sha256-ec7b9f5edabec15412e2e4d58268d0e40d900c57bb59fa62a09e41b991d0f596?context=explore) on https://hub.docker.com/r/opensearchstaging/opensearch/tags that includes the asynchronous-search plugin as well and replace it in CI (might need to try locally).

I tried to use newer tag in 3.0 build, still same error. Can you check it? I guess I made a mistake somewhere
image

**Comment by dblock** [comment]
@Tokesh I don't think it's a mistake, this likely means that asynchronous-search doesn't build in 3.0, and isn't included in any of these builds on docker hub 

@peterzhuamazon Can you help? How can one know which 3.0 build has asynchronous-search?

**Comment by dblock** [comment]
@Tokesh For now add `< 3.0` to skip this test so we can merge this code?

**Comment by Tokesh** [comment]
> @Tokesh I don't think it's a mistake, this likely means that asynchronous-search doesn't build in 3.0, and isn't included in any of these builds on docker hub
> 
> @peterzhuamazon Can you help? How can one know which 3.0 build has asynchronous-search?

Actually, i made a quick research about this situation.
I found `async_search` in OS 3.0 build [here](https://github.com/opensearch-project/asynchronous-search/issues/502)
But I don't know why it doesn't work now.

**Comment by Tokesh** [comment]
> @Tokesh For now add `< 3.0` to skip this test so we can merge this code?

Ready for your review :)

**Comment by dblock** [comment]
> Actually, i made a quick research about this situation. I found `async_search` in OS 3.0 build [here](https://github.com/opensearch-project/asynchronous-search/issues/502) But I don't know why it doesn't work now. I have hypothesis, that it shouldn't work like this

Will you take a look at this? It should work in 3.0 the same as in 2.x, or it's a bug.

**Review by dblock — APPROVED** [review]

**Comment by Tokesh** [comment]
> > Actually, i made a quick research about this situation. I found `async_search` in OS 3.0 build [here](https://github.com/opensearch-project/asynchronous-search/issues/502) But I don't know why it doesn't work now. I have hypothesis, that it shouldn't work like this
> 
> Will you please take a look at this? At least open a bug. It should work in 3.0 the same as in 2.x, or it's a bug.

Yeap sure! I'm already opened issue in async-search repo.

**Comment by dblock** [comment]
> Yeap sure! I'm already opened issue in async-search repo.

Thanks. That repo could use help if you want to try fixing it!

**Pull request merged**